### ISSUE TITLE
Question: How can I obtain day of year (1-366) from a serial date

### ISSUE THREAD

**Issue by dibyendumajumdar** [HowardHinnant/date#169]
Hi

I need the ability to calculate the day of year as an offset from 1 to 366 - given a date. How can I obtain this? Apologies if this is already available as an API or has been answered - I could not find the relevant information.

I see that there is a doy variable that is used when converting from serial date to the date components. But I am not sure that this is what I am looking for.

Thanks and Regards
Dibyendu

**Comment by dibyendumajumdar** [comment]
I suppose the answer is this:?

https://github.com/HowardHinnant/date/wiki/Examples-and-Recipes#year_day

Didn't spot that before - but I would like to avoid calculating another date if possible.

**Comment by HowardHinnant** [comment]
Yes, that's correct.  Here's how I would do it (same idea, more compact syntax):
````
date::days
get_day_offset(date::year_month_day ymd)
{
    using namespace date;
    return sys_days{ymd} - sys_days{(ymd.year()-years{1})/dec/31};
}
````
It would be exercised like this:
````
#include <iostream>

int
main()
{
    using namespace std::chrono;
    using namespace date;
    std::cout << get_day_offset(floor<days>(system_clock::now())) << '\n';
}
````
which just output for me:
````
183[86400]s
````
(183 days).

Note that I chose the epoch such that Jan 1 is day 1, not day 0.

**Comment by dibyendumajumdar** [comment]
Thank you. Is there any way to get the ordinal date as a byproduct of the function that computes the y, m, d fields? I saw a doy variable there - but although the comment says it is [0-365] - I saw that is not actually the value I was looking for. 

Regards

**Comment by HowardHinnant** [comment]
No, sorry, there is not.  `doy` contains the days since March 1.

**Comment by HowardHinnant** [comment]
Fwiw, I've just completed a performance analysis of the `get_day_offset` algorithm above (on clang -O3 macOS).  My best estimate is that you will be hard put to develop a faster algorithm.  Yes, I believe a faster algorithm could be done.  But it would be complicated and unreadable.

I've looked at the assembly of `get_day_offset`:  It is 86 non-branching x86_64 assembly instructions.  Furthermore (and probably more importantly), there is _no_ distant memory access.  It is all operating on registers and immediates. That means no cache misses.

Timings on something this short are difficult.  But I attempted a loop over one month, being careful that the loop did not optimize out what I was trying to measure by use of `volatile`, and inspecting the assembly.  My best guess on a 3 yr old consumer laptop based on a 2.4 GHz Intel Core i5 is that this function will take somewhere between 22ns and 45ns to execute.

Any algorithm to do this that involves loops and/or lookup tables won't be able to compete.

A special purpose algorithm that takes apart [`days_from_civil`](http://howardhinnant.github.io/date_algorithms.html#days_from_civil) will be able to beat it, but will be hard to read, and harder to test with certainty that it is correct.  And by looking at the instruction count of these two functions, I estimate the optimized algorithm would be in the neighborhood of 13ns to 27ns -- faster, but not twice as fast.

Though I have not measured, my best guess is that any other library to perform the functionality of `get_day_offset`, besides one that carefully dissects [`days_from_civil`](http://howardhinnant.github.io/date_algorithms.html#days_from_civil), is going to be iterative and branching, and thus much slower than `get_day_offset`.
